File: speech_to_text/java/real_time_transcription/realtime/src/main/java/com/soniox/example/realtime/Realtime.java
package com.soniox.example.realtime;

import java.io.*;
import java.net.http.HttpClient;
import java.net.http.WebSocket;
import java.net.URI;
import java.nio.*;
import java.util.concurrent.*;

import org.json.JSONObject;

public class Realtime {
    public static void main(String[] args) {
        String apiKey = System.getenv("SONIOX_API_KEY");

        ExecutorService executor = Executors.newCachedThreadPool();

        HttpClient client = HttpClient.newBuilder().executor(executor).build();

        URI wsUri = URI.create("wss://stt-rt.soniox.com/transcribe-websocket");

        CountDownLatch latch = new CountDownLatch(1);

        // Connect to WebSocket API
        System.out.println("Opening WebSocket connection...");

        WebSocket webSocket = client.newWebSocketBuilder().buildAsync(wsUri, new WebSocket.Listener() {
            @Override
            public void onOpen(WebSocket webSocket) {
                // Send start request

                JSONObject startRequest = new JSONObject();
                startRequest.put("api_key", apiKey);
                startRequest.put("audio_format", "pcm_s16le");
                startRequest.put("sample_rate", 16000);
                startRequest.put("num_channels", 1);
                startRequest.put("model", "stt-rt-preview");

                webSocket.sendText(startRequest.toString(), true);

                // Start send audio thread

                Thread sendAudioThread = new Thread(() -> {
                    // Send audio data

                    try (InputStream inputStream = new FileInputStream("coffee_shop.pcm_s16le")) {
                        System.out.println("Transcription started.");

                        try {
                            byte[] buffer = new byte[3840];
                            int bytesRead;

                            while ((bytesRead = inputStream.read(buffer)) != -1) {
                                webSocket.sendBinary(ByteBuffer.wrap(buffer, 0, bytesRead), true).join();
                                // Sleep for 120 ms
                                TimeUnit.MILLISECONDS.sleep(120);
                            }

                            // Send send of file

                            webSocket.sendText("", true).join();
                        } catch (IOException | InterruptedException | CompletionException e) {}
                    } catch (IOException e) {
                        e.printStackTrace();
                        System.exit(1);
                    }
                });
                executor.submit(sendAudioThread);

                WebSocket.Listener.super.onOpen(webSocket);
            }

            @Override
            public CompletionStage<?> onText(WebSocket webSocket, CharSequence data, boolean last) {
                // Receive text

                JSONObject response = new JSONObject(data.toString());

                // Handle any possible transcription error
                if (response.opt("error_code") != null)
                {
                    System.out.println("\nError: " + response.getInt("error_code") + ": " + response.getString("error_message"));
                    webSocket.sendClose(WebSocket.NORMAL_CLOSURE, "");
                } else {
                    // Process tokens
                    response.getJSONArray("tokens").forEach(t -> {
                        JSONObject token = (JSONObject) t;
                        if (token.getString("text") != null) {
                            System.out.print(token.getString("text"));
                            System.out.flush();
                        }
                    });

                    if (response.opt("finished") != null && response.getBoolean("finished")) {
                        // Close the WebSocket
                        System.out.println("\nTranscription done.");
                        webSocket.sendClose(WebSocket.NORMAL_CLOSURE, "");
                    }
                }

                return WebSocket.Listener.super.onText(webSocket, data, last);
            }

            @Override
            public CompletionStage<?> onClose(WebSocket webSocket, int statusCode, String reason) {
                System.out.println("");

                if (!reason.equals("")) {
                    System.out.println("Close error: " + reason);
                }

                latch.countDown();

                return WebSocket.Listener.super.onClose(webSocket, statusCode, reason);
            }

            @Override
            public void onError(WebSocket webSocket, Throwable error) {
                System.out.println("");

                error.printStackTrace();

                latch.countDown();
            }
        }).join();

        try {
            latch.await();
        } catch (InterruptedException e) {}

        executor.shutdownNow();
    }
}

File: speech_to_text/php/real_time_transcription/realtime/realtime.php
<?php

require "vendor/autoload.php";

use Ratchet\Client\WebSocket;
use Ratchet\Client\Connector;
use Ratchet\RFC6455\Messaging\Frame;
use React\EventLoop\Factory;
use React\Socket\ConnectionException;
use React\Stream\ReadableResourceStream;

$api_key = getenv("SONIOX_API_KEY");
$ws_uri = "wss://stt-rt.soniox.com/transcribe-websocket";
$file_to_stream = "coffee_shop.pcm_s16le";

$loop = Factory::create();

$connector = new Connector($loop);

// Connect to WebSocket API
echo "Opening WebSocket connection..." . PHP_EOL;
$connector($ws_uri)->then(
    function (WebSocket $ws) use ($api_key, $loop, $file_to_stream) {
        // Send start request
        $ws->send(
            json_encode([
                "api_key" => $api_key,
                "audio_format" => "pcm_s16le",
                "sample_rate" => 16000,
                "num_channels" => 1,
                "model" => "stt-rt-preview"
            ])
        );

        $audio_file = fopen($file_to_stream, "rb");
        if (!$audio_file) {
            die("Failed to open audio file." . PHP_EOL);
        }

        echo "Transcription started." . PHP_EOL;

        // Read and send audio data from file over WebSocket connection every 120 ms
        $timer = $loop->addPeriodicTimer(0.12, function ($timer) use (
            $loop,
            $ws,
            $audio_file
        ) {
            $data = fread($audio_file, 3840);

            if (feof($audio_file) || empty($data)) {
                // Send end of file

                $ws->send("");
                $loop->cancelTimer($timer);
                return;
            }

            $ws->send(new Frame($data, true, Frame::OP_BINARY));
        });

        $ws->on("message", function ($msg) {
            // Receive text
            $res = json_decode($msg, true);

            if (isset($res["error_code"])) {
                echo PHP_EOL . "Error: " . $res["error_code"] . " " . $res["error_message"] . PHP_EOL;
                return;
            }

            foreach ($res["tokens"] as $token) {
                echo $token["text"];
            }

            if (array_key_exists("finished", $res) && $res["finished"]) {
                echo PHP_EOL . "Transcription done." . PHP_EOL;
            }
        });

        $ws->on("close", function ($code = null, $reason = null) use (
            $loop,
            $timer
        ) {
            $loop->cancelTimer($timer);

            if ($reason != "") {
                echo "Close error: $code - $reason" . PHP_EOL;
            }
        });
    },
    function (ConnectionException $e) {
        echo "Could not connect: {$e->getMessage()}" . PHP_EOL;
    }
);

$loop->run();

?>
